Add generic EVP method fetcher

levitte opened the pull request:
This is an interface between Core dispatch table fetching and
EVP_{method}_fetch().  All that's needed from the diverse method
fetchers are the functions to create a method structure from a
dispatch table, a function that ups the method reference counter and a
function to free the method (in case of failure).

This routine is internal to the EVP API andis therefore only made
accessible within crypto/evp, by including evp_locl.h

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [ ] tests are added or updated

levitte committed:
Add generic EVP method fetcher

This is an interface between Core dispatch table fetching and
EVP_{method}_fetch().  All that's needed from the diverse method
fetchers are the functions to create a method structure from a
dispatch table, a function that ups the method reference counter and a
function to free the method (in case of failure).

This routine is internal to the EVP API andis therefore only made
accessible within crypto/evp, by including evp_locl.h

levitte commented:
To build, this needs #8286, #8287 and #8340.  It's here to show how internal API specific methods can get created from provider dispatch tables through the core fetcher from #8340.

paulidale commented on crypto/evp/evp_fetch.c:
```diff
@@ -0,0 +1,215 @@
+#include "internal/asn1_int.h"
+#include "internal/property.h"
+#include "internal/core.h"
+#include "internal/evp_int.h"    /* evp_locl.h needs it */
```
Would it make sense to include this from `evp_locl.h` if it is needed there?

paulidale commented on crypto/evp/evp_fetch.c:
```diff
@@ -0,0 +1,215 @@
+static int default_method_store_index = -1;
+struct default_method_store_st {
+    OSSL_METHOD_STORE *store;
+    CRYPTO_RWLOCK *lock;
```
The store contains a lock -- this seems like duplication.
I'd suggest leaving both in place for the moment and removing one later if it is possible.

levitte replied:
Oh right, I can just call the `ossl_method_store` functions without a surrounding lock.  I had forgotten

levitte replied:
So errr, contrary to your suggestion, I removed the lock here.

paulidale replied:
No problem.  I wasn't sure if you'd need a lock at this level or not.  It seems not.

levitte replied:
No...  this code was much designed around earlier incantations of your store code, I think there wasn't a general lock back then.

paulidale commented:
CI errors are relevant.

levitte commented:
@paulidale, in https://github.com/openssl/openssl/pull/8341#issuecomment-467359497, I mention that this needs #8286, #8287 and #8340.  Only #8286 has yet made it into master.

paulidale commented:
Okay, I'll look at them too.

slontis commented on doc/internal/man3/evp_generic_fetch.pod:
```diff
@@ -0,0 +1,232 @@
+    /* typedef struct evp_foo_st EVP_FOO_CTX */
+    struct evpn_foo_ctx_st {
+        const EVP_FOO *foo;
+        void *provctx;
```
I find this name a bit confusing

levitte replied:
I'll add a comment

slontis commented on doc/internal/man3/evp_generic_fetch.pod:
```diff
@@ -0,0 +1,232 @@
+    int EVP_FooInit(EVP_FOO_CTX *c, const EVP_FOO *foo)
+    {
+        int ok = 1;
+        size_t ivl = iv == NULL ? 0 : 9999;
```
??

levitte replied:
Remnant from experimental code.  I will remove

slontis commented on doc/internal/man3/evp_generic_fetch.pod:
```diff
@@ -0,0 +1,232 @@
+        OSSL_OP_foo_operate_fn *operate;
+        OSSL_OP_foo_cleanctx_fn *cleanctx;
+        OSSL_OP_foo_freectx_fn *freectx;
+    };
```
refcnt?

levitte replied:
Ah, good point

slontis commented on crypto/evp/evp_fetch.c:
```diff
@@ -0,0 +1,181 @@
+    if (nid != NID_undef
+        && ossl_method_store_cache_get(NULL, nid, properties,
+                                       (void **)&method)) {
+        upref_method(method);
```
does the return value need to be checked?

levitte commented:
I've rebased this on top of a fresh master.  I'll go through any remaining comments tomorrow

mattcaswell commented on crypto/evp/evp_fetch.c:
```diff
@@ -0,0 +1,182 @@
+        tmpobj.length = 0;
+        tmpobj.data = NULL;
+
+        methdata->nid = OBJ_add_object(&tmpobj);
```
Do we need some locking here?

levitte replied:
(and for anyone listening, PR welcome!)

mattcaswell replied:
We should raise an issue for this

mattcaswell commented on crypto/evp/evp_fetch.c:
```diff
@@ -0,0 +1,182 @@
+                               void *data)
+{
+    struct method_data_st *methdata = data;
+    int ret = 0;
```
ret is unused

levitte replied:
Fixed

levitte commented on crypto/evp/evp_fetch.c:
```diff
@@ -0,0 +1,182 @@
+
+    if (nid != NID_undef
+        && ossl_method_store_cache_get(NULL, nid, properties,
+                                       (void **)&method)) {
```
Because compilers are "helpfully" warning about the type loss otherwise...  it's the same kind of helpfulness as when you try to pass a `char **` to a `const char **` parameter.  It's really annoying, but there you go...

mattcaswell replied:
I must be being stupid...what type loss are you referring to? I don't see any.

levitte replied:
Looks like you're right...  this must be an artefact from older versions of the code that didn't get cleaned away.  So, er, thanks :-)

levitte replied:
Fixed

levitte committed:
fixup! Add generic EVP method fetcher

levitte committed:
fixup! Add generic EVP method fetcher

levitte commented:
Please note, btw, that this won't work without #8340...

levitte commented:
#8340 is now merged, and this has been rebased on top.

mattcaswell commented:
Travis errors look relevant:
````
crypto/evp/evp_fetch.c: In function 'evp_generic_fetch':
crypto/evp/evp_fetch.c:168:9: error: initializer element is not computable at load time [-Werror]
         };
         ^
cc1: all warnings being treated as errors
make[1]: *** [crypto/evp/libcrypto-shlib-evp_fetch.o] Error 1
make[1]: Leaving directory `/home/travis/build/openssl/openssl'
make: *** [build_programs] Error 2
````

levitte committed:
Replumbing: pass callback data to the algo destructor too

All relevant OSSL_METHOD_CONSTRUCT_METHOD callbacks got the callback
data passed to them, except 'destruct'.  There's no reason why it
shouldn't get that pointer passed, so we make a small adjustment.

levitte committed:
fixup! Add generic EVP method fetcher

levitte commented:
Saw it, and discovered a small thing missing in #8340.  Corrected with 
837f046

levitte commented:
CIs are happy now

mattcaswell commented on doc/internal/man3/evp_generic_fetch.pod:
```diff
@@ -0,0 +1,232 @@
+                                            unsigned char *out, size_t *out_l))
+                                            unsigned char *in, size_t in_l))
+    OSSL_core_make_func(void,OP_foo_cleanctx,(void *vctx))
+    OSSL_core_make_func(void,OP_foo_freectx,(void *vctx))
```
s/OSSL_core_make_func/OSSL_CORE_MAKE_FUNC/

levitte replied:
Fixed

mattcaswell commented on doc/internal/man3/evp_generic_fetch.pod:
```diff
@@ -0,0 +1,232 @@
+    OSSL_core_make_func(int,OP_foo_init,(void *vctx))
+    OSSL_core_make_func(int,OP_foo_operate,(void *vctx,
+                                            unsigned char *out, size_t *out_l))
+                                            unsigned char *in, size_t in_l))
```
mis-matched parens here

levitte replied:
Fixed

levitte committed:
Replumbing: better reference counter control in ossl_method_construct()

Fully assume that the method constructors use reference counting.
Otherwise, we may leak memory, or loose track and do a double free.

levitte committed:
fixup! Add generic EVP method fetcher

levitte committed:
fixup! Add generic EVP method fetcher

levitte commented:
The work done today was to correct some refcounting that wasn't quite right.

mattcaswell commented on crypto/core_fetch.c:
```diff
@@ -39,25 +39,33 @@ static int ossl_method_construct_this(OSSL_PROVIDER *provider, void *cbdata)
+         * Note regarding putting the method in stores:
+         *
+         * we don't need to care if it actually got in or not here.
+         * If it didn't get it, it will simply not be available when
```
s/get it/get in/

levitte replied:
Fixed

mattcaswell reviewed (approved):
Approved assuming the minor typo is fixed. I also think an issue should be raised for the OBJ locking stuff before it gets forgotten about.

levitte commented:
Merged.

Thanks for the reviews!

c13d2ab439a9dcbbf22ef85a00603142b0a37779 Add generic EVP method fetcher
a383083194b882a904ae66fcf74ebc348602407c Replumbing: better reference counter control in ossl_method_construct()
7bb19a0f950dd87607133d526e31a083f35921fd Replumbing: pass callback data to the algo destructor too
